models/items.js

```diff
@@ -0,0 +1,225 @@
+const firestore = require("../utils/firestore");
+
+const itemModel = firestore.collection("items");
+const tagModel = firestore.collection("tags");
+const levelModel = firestore.collection("levels");
+
+/**
+ *
+ * @param itemData { Object }: Data of the item, item type and Array of pairs of tagid and levelid
+ * @returns {Promise<tagModel|String>}
+ */
+
+const addTagsToItem = async (itemData) => {
+  try {
+    const { itemid, itemType, tagPayload } = itemData;
+    const batch = firestore.batch();
+    for (const tag of tagPayload) {
+      const tagData = await tagModel.doc(tag.tagid).get();
+      const levelData = await levelModel.doc(tag.levelid).get();
+      const itemTag = {
+        itemid,
+        itemtype: itemType.toUpperCase(),
+        tagid: tag.tagid,
+        tagname: tagData.data().name,
+        tagtype: tagData.data().type,
+        levelid: tag.levelid,
+        levelname: levelData.data().name,
+        levelnumber: levelData.data().levelnumber,
+      };
+      const docid = itemModel.doc();
+      batch.set(docid, itemTag);
+    }
+    await batch.commit();
+    return { itemid };
+  } catch (err) {
+    logger.error("Error in creating Item", err);
```

Comment:
If you using commit in saving data in batch, if an error occurs, it should rollback already added data,
The rollback part is missing. If data is a rollback, add a finally block or rollback in the catch part.

Reply:
https://stackoverflow.com/questions/60864067/firestore-batch-write-rollback

This article states that firestore doesnt have the rollback feature